components/script/dom/htmlimageelement.rs

```diff
@@ -271,106 +268,360 @@ impl HTMLImageElement {
+                        // Step 12.2
+                        if parsed_url == *url {
+                            // 12.3 abort pending request
+                            let mut pending_request = self.pending_request.borrow_mut();
+                            pending_request.image = None;
+                            pending_request.parsed_url = None;
+                            LoadBlocker::terminate(&mut pending_request.blocker);
+                            // TODO: queue a task to restart animation, if restart-animation is set
+                            return
                         }
+                        self.init_pending_request(&url, &src);
+                        self.image_request.set(ImageRequestPhase::Pending);
+                        self.fetch_image(&url);
+                    },
+                    (_, State::Broken) | (_, State::Unavailable) => {
+                        // Step 12.5
+                        current_request.parsed_url = Some(url.clone());
+                        current_request.source_url = Some(src.clone());
+                        current_request.image = None;
+                        current_request.metadata = None;
+                        let document = document_from_node(self);
+                        current_request.blocker = Some(LoadBlocker::new(&*document, LoadType::Image(url.clone())));
+                        self.fetch_image(&url);
+                    },
+                    (_, _) => {
+                        // step 12.6
+                        self.init_pending_request(&url, &src);
+                        self.image_request.set(ImageRequestPhase::Pending);
+                        self.fetch_image(&url);
+                    },
+                }
+            }
+        }
+    }
+
+    /// Step 8-12 of html.spec.whatwg.org/multipage/#update-the-image-data
+    fn update_the_image_data_sync_steps(&self) {
+        let document = document_from_node(self);
+        // Step 8
+        // TODO: take pixel density into account
+        match self.select_image_source() {
+            Some(src) => {
+                // Step 10
+                self.dispatch_progress_event(atom!("loadstart"), 0, None);
+                // Step 11
+                let base_url = document.base_url();
+                let parsed_url = base_url.join(&src);
+                match parsed_url {
+                    Ok(url) => {
+                         // Step 12
+                        self.prepare_image_request(&url, &src);
+                    },
+                    Err(_) => {
+                        // Step 11.1-11.5
+                        self.abort_request(State::Broken, ImageRequestPhase::Current);
+                        self.abort_request(State::Broken, ImageRequestPhase::Pending);
+                        self.set_current_request_url_to_selected(src);
+                        self.dispatch_event(atom!("error"));
+                        self.dispatch_event(atom!("loadend"));
```

Comment:
`dispatch_event` also does the dispatching by scheduling a task to do so, I think that's what you mean? Same as with https://github.com/servo/servo/pull/16238/files#r114092939